specs/protocol/block_header.md

```diff
@@ -0,0 +1,24 @@
+# Block Header
+
+## Application Header
+
+name                 | type       | description
+---------------------|------------|-----------------------------------------------------------------------------------------------------
+`height`             | `uin64`    | Height of this block.
+`timestamp`          | `uin64`    | Time this block was created, in [TAI64](https://cr.yp.to/libtai/tai64.html) format.
+`txCount`            | `uint64`   | Number of [transaction](./tx_format.md#transaction)s in this block.
+`txRoot`             | `byte[32]` | [Merkle root](./cryptographic_primitives.md#binary-merkle-tree) of [transaction](./tx_format.md#transaction)s in this block.
+`outputMessagesCount`|`uint64`| Number of [output message](./tx_format.md#outputmessage)s in this block.
+`outputMessagesRoot` | `byte[32]` | [Merkle root](./cryptographic_primitives.md#binary-merkle-tree) of [output message](./tx_format.md#outputmessage)s in this block.
```

Comment:
Do we need messages as separate from the transactions here? @pixelcircuits

Reply:
seems like they should be derivable from the transaction root?

Reply:
It is, yes. But if validators don't sign over it then there's no compact way of posting such a root in an authenticated way to an Ethereum contract.

Reply:
My thinking was for validators to sign over the entire state of output messages set per block, that way we only have to periodically commit to Ethereum rather than commit a hash/root for each and every block.

Reply:
> It is, yes. But if validators don't sign over it then there's no compact way of posting such a root in an authenticated way to an Ethereum contract.

Wouldn't this apply to all the data posted to ethereum, and not just message outputs? How does splitting messages out into a separate field address this issue?

> My thinking was for validators to sign over the entire state of output messages set per block, that way we only have to periodically commit to Ethereum rather than commit a hash/root for each and every block.

Let's assume the validators signed the current block hash MMR root, and the commit also includes a proof that the current block is linked to the last committed block. Seems like that may be sufficient to prove any of the other fields in previous blocks that may or may not have been committed to eth? 

This would also require a commit every time the validator staking set is modified, but that could be amortized using epochs. I.e. staking adjustments are batched up and only applied every epoch (some arbitrary duration like ~1000 blocks).

Reply:
We _could_ just have the proof travel down the chain to a certain block first. I'm not crazy enthusiastic about this idea though, since the proof size would grow almost linearly depending on when the message was outputted vs the next commit to Ethereum. I think this would end up in users trying to time txs to right before the state gets committed which could cause regular periodic spikes in L2 gas fees.

Reply:
How about for now, we remove the output messages from being their own thing in the "header" as they can be proven/derived from the tx root. Then when we figure out the concept of "epochs" (which will be needed to freeze the validator set to prevent having to check in every single block to Ethereum) we have it include committing a rolled up merkle tree of all message outputs in the past epoch

Reply:
> How about for now, we remove the output messages from being their own thing in the "header" as they can be proven/derived from the tx root.

That would require users that want to consume an output message to provide an entire transaction to Ethereum to be parsed. If the output messages are in their own tree, then only the output message itself needs to be proven.

Reply:
Having the message merkle root separated from txroot would help to limit the amount of leaf data that needs to be proved by users. Since there's not really much use case to prove anything else in a tx currently.